src/bmson.rs

```diff
@@ -382,35 +380,41 @@ where
             E: Error,
         {
             match v {
-                0 => NonZeroU64::new(default_resolution())
-                    .ok_or_else(|| E::custom("default_resolution should be non-zero")),
-                v => NonZeroU64::new(v)
-                    .ok_or_else(|| E::custom(format!("Invalid resolution value: {}", v))),
+                0 => Ok(default_resolution_nonzero()),
+                v => Ok(NonZeroU64::new(v)
+                    .expect("NonZeroU64::new should not fail for non-zero u64 value")),
             }
         }
 
         fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
         where
             E: Error,
         {
-            if v == 0.0 {
-                NonZeroU64::new(default_resolution())
-                    .ok_or_else(|| E::custom("default_resolution should be non-zero"))
-            } else if v.fract() == 0.0 && v > 0.0 {
-                // Only accept whole positive numbers
-                let as_u64 = v as u64;
-                if (as_u64 as f64) == v {
-                    NonZeroU64::new(as_u64)
-                        .ok_or_else(|| E::custom(format!("Invalid resolution value: {}", v)))
-                } else {
-                    Err(E::custom(format!("Resolution value too large: {}", v)))
-                }
-            } else {
-                Err(E::custom(format!(
-                    "Resolution must be a positive whole number, got: {}",
-                    v
-                )))
+            // Bmson spec: if negative, take the absolute value
+            let av = v.abs();
+            if !av.is_finite() {
+                return Err(E::custom("Resolution must be a finite number"));
             }
+
+            // Accept numbers that round to an integer
+            let rounded = av.round();
+            assert!(
+                rounded >= 0.0,
+                "Rounded value is negative: {} -> {}",
+                v,
+                rounded
+            );
+            if rounded == 0.0 {
+                return Ok(default_resolution_nonzero());
+            }
+
+            // Ensure the rounded value fits in u64 exactly
+            if rounded > (u64::MAX as f64) {
+                return Err(E::custom(format!("Resolution value too large: {}", v)));
+            }
```

Comment:
According to [WebIDL definition of Bmson](https://bmson-spec.readthedocs.io/en/master/doc/index.html#format-overview) and other descriptions, `resolution` is defined as an `unsigned long`. So I think non-integer (non-zero `av.fract()`) should be an error. Or does some Bmson player round them?